File: List/206.反转链表.cpp
/*
 * @lc app=leetcode.cn id=206 lang=cpp
 *
 * [206] 反转链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
public:
    ListNode *reverseList(ListNode *head)
    {
        // 记录上一个元素，也就是最终返回的头指针
        // 这个题的做法是将链表的指针反转方向即可
        ListNode *pre = nullptr;
        ListNode *current = head;

        while (current != nullptr)
        {
            // 临时保存下一个元素
            ListNode *temp = current->next;

            // 当前元素的指针指向上一个元素
            current->next = pre;
            // 记录当前元素，下一次遍历时候使用
            pre = current;
            // 继续遍历下一个元素
            current = temp;
        }

        return pre;
    }
};
// @lc code=end

File: List/206.反转链表.ts
/*
 * @lc app=leetcode.cn id=206 lang=typescript
 *
 * [206] 反转链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function reverseList(head: ListNode | null): ListNode | null {
  if (!head || !head.next) {
    return head
  }

  // pre记录上一个节点，所以，这里以null开始
  let pre = null
  let current = head
  let firstNode = null

  while (current && current.next) {
    // 新建一个节点作为当前的头结
    firstNode = new ListNode(current.val, pre)
    // 头节点（也就是当前节点）作为下一次遍历的pre节点
    pre = firstNode
    // 继续下一次遍历
    current = current.next
  }

  // 遍历到最后一个节点跳出循环，手动赋值最后一个节点到表头
  head = new ListNode(current.val, firstNode)

  return head
};
// @lc code=end

